db/test.cc

```diff
@@ -0,0 +1,126 @@
+
+#define BRICKS_MOCK_TIME
+
+#include "../../Bricks/port.h"
+
+#include <string>
+
+#include "../db.h"
+
+#include "../../Bricks/dflags/dflags.h"
+#include "../../Bricks/3party/gtest/gtest-main-with-dflags.h"
+
+using bricks::Singleton;
+using bricks::strings::Printf;
+
+DEFINE_int32(test_port, 8091, "Local port to use for the test.");
+
+struct ListenOnTestPort {
+  ListenOnTestPort() {
+    HTTP(FLAGS_test_port).Register("/", [](Request r) { r("I'm listening, baby.\n"); });
+  }
+};
+
+struct UnitTestStorageListener {
+  std::atomic_size_t n;
+  std::string data;
+  UnitTestStorageListener() : n(0u) {}
+  inline bool Entry(const std::unique_ptr<schema::Record>& entry) {
+    data += JSON(entry, "record") + "\n";
+    ++n;
+    return true;
+  }
+  inline void Terminate() {
+    data += "DONE\n";
+    ++n;
+  }
+};
+
+TEST(AgreeDisagreeDemo, EndpointsAndScope) {
+  const std::string url_prefix = Printf("http://localhost:%d", FLAGS_test_port);
+  // `/test1` is inactive. Ensure that an HTTP server is listening on the port though.
+  Singleton<ListenOnTestPort>();
+  EXPECT_EQ(404, static_cast<int>(HTTP(GET(url_prefix + "/test1")).code));
+  {
+    db::Storage storage(FLAGS_test_port, "test1");
+    // `/test1` is available in the scope of `Storage("test1");`.
+    EXPECT_EQ(200, static_cast<int>(HTTP(GET(url_prefix + "/test1")).code));
+  }
+  // `/test1` is inactive again as `Storage("test1");` gets out of scope.
+  EXPECT_EQ(404, static_cast<int>(HTTP(GET(url_prefix + "/test1")).code));
+}
+
+TEST(AgreeDisagreeDemo, Questions) {
+  db::Storage storage(FLAGS_test_port, "test2");
+
+  UnitTestStorageListener listener;
+  auto listener_scope = storage.Subscribe(listener);
+
+  const std::string url_prefix = Printf("http://localhost:%d", FLAGS_test_port);
+  // The question with QID=1 does not exist.
+  EXPECT_EQ(404, static_cast<int>(HTTP(GET(url_prefix + "/test2/q?qid=1")).code));
+  // A question can be added and gets a QID of 1.
+  EXPECT_EQ(0u, listener.n);
+  bricks::time::SetNow(bricks::time::EPOCH_MILLISECONDS(1001));
+  const auto added = HTTP(POST(url_prefix + "/test2/q?text=Why%3F"));
+  EXPECT_EQ(200, static_cast<int>(added.code));
+  EXPECT_EQ("{\"question\":{\"ms\":1001,\"qid\":1,\"text\":\"Why?\"}}\n", added.body);
+  // A new question with the same text can not be added.
+  EXPECT_EQ(400, static_cast<int>(HTTP(POST(url_prefix + "/test2/q?text=Why%3F")).code));
+  // A question with QID of 1 can be retrieved now.
+  const auto retrieved = HTTP(GET(url_prefix + "/test2/q?qid=1"));
+  EXPECT_EQ(200, static_cast<int>(retrieved.code));
+  EXPECT_EQ("{\"value0\":{\"ms\":1001,\"qid\":1,\"text\":\"Why?\"}}\n", retrieved.body);
+
+  // Ensure that the question is processed as it reaches the listener stream.
+  while (listener.n != 1) {
+    ;  // Spin lock;
+  }
+
+  EXPECT_EQ(
+      "{\"record\":{\"polymorphic_id\":2147483649,\"polymorphic_name\":\"Q\",\"ptr_wrapper\":"
```

Comment:
Which part?

Based on our research, this format is consistent (and I am just double-checking with Cereal 1.1.0 as we speak.)